SeQuant/core/binary_node.hpp

```diff
@@ -9,6 +10,77 @@
 
 namespace sequant {
 
+template <typename>
+class FullBinaryNode;
+
+/// Full-binary node visit orders.
+
+/// Visit children nodes followed by parent node.
+/// Left will be visited before right.
+struct PreOrder {};
+
+/// Visit parent node followed by children nodes.
+/// Left will be visited before right.
+struct PostOrder {};
+
+/// Visit left node then parent node followed by right node.
+struct InOrder {};
```

Comment:
Might it make sense to group these into an enum?
```suggestion
enum class IterationOrder {
  Pre,
  Post,
  In,
};
```

Reply:
Then we cannot use tag dispatch. Which I intend to use.

Reply:
Ah, I see.

One could `if constexpr` on the enum value and call the desired overloads, but I guess that could get clumsy :eyes: